Python/optimizer.c

```diff
@@ -0,0 +1,248 @@
+
+#include "Python.h"
+#include "opcode.h"
+#include "pycore_interp.h"
+#include "pycore_opcode.h"
+#include "pycore_pystate.h"
+#include "cpython/optimizer.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+
+static int32_t
+get_next_free_in_executor_array(PyCodeObject *code)
+{
+    _PyExecutorArray *old = code->co_executors;
+    int size = 0;
+    int capacity = 0;
+    if (old != NULL) {
+        size = old->size;
+        capacity = old->capacity;
+        if (capacity >= 256) {
+            PyErr_Format(PyExc_SystemError,
+                        "too many executors for code object");
+            return -1;
+        }
+    }
+    for (uintptr_t i = 0; i < size; i++) {
+        if (old->executors[i] == NULL) {
+            return i;
+        }
+    }
+    if (size >= capacity) {
+        int new_capacity = capacity ? capacity * 2 : 4;
+        _PyExecutorArray *new = PyMem_Malloc(offsetof(_PyExecutorArray, executors) +
+            new_capacity * sizeof(_PyExecutorObject *));
+        if (new == NULL) {
+            PyErr_NoMemory();
+            return -1;
+        }
+        code->co_executors = new;
+        new->capacity = new_capacity;
+        for (uintptr_t i = 0; i < size; i++) {
+            new->executors[i] = old->executors[i];
+        }
```

Comment:
This chunk could potentially be replaced with a `PyMem_Realloc`. That essentially does a `malloc` + `memcpy`.